quantum/rgblight.c

```diff
@@ -596,11 +604,69 @@ void rgblight_sethsv_master(uint8_t hue, uint8_t sat, uint8_t val) { rgblight_se
 void rgblight_sethsv_slave(uint8_t hue, uint8_t sat, uint8_t val) { rgblight_sethsv_range(hue, sat, val, (uint8_t)RGBLED_NUM / 2, (uint8_t)RGBLED_NUM); }
 #endif  // ifndef RGBLIGHT_SPLIT
 
+#ifdef RGBLIGHT_LAYERS
+static bool rgblight_is_static_mode(void) {
+    uint8_t mode = rgblight_status.base_mode;
+    bool is_static = 
+#   ifdef RGBLIGHT_EFFECT_STATIC_GRADIENT
+                    mode == RGBLIGHT_MODE_STATIC_GRADIENT ||
+#   endif
+                    mode == RGBLIGHT_MODE_STATIC_LIGHT;
+    return is_static;
+}
+
+void rgblight_set_layer_state(uint8_t layer, bool enabled) {
+    uint8_t mask = 1 << layer;
+    if (enabled) {
+        rgblight_status.enabled_layer_mask |= mask;
+    } else {
+        rgblight_status.enabled_layer_mask &= ~mask;
+    }
+#    ifdef RGBLIGHT_SPLIT
+    rgblight_status.change_flags |= RGBLIGHT_STATUS_CHANGE_LAYERS;
+#    endif
+    // Static modes don't have a ticker running to update the LEDs
+    if (rgblight_is_static_mode()) {
+        rgblight_mode_noeeprom(rgblight_get_mode());
+    }
+}
+
+bool rgblight_get_layer_state(uint8_t layer) {
+    uint8_t mask = 1 << layer;
+    return (rgblight_status.enabled_layer_mask & mask) > 0;
+}
+#endif
+
 #ifndef RGBLIGHT_CUSTOM_DRIVER
 void rgblight_set(void) {
     LED_TYPE *start_led;
     uint16_t  num_leds = clipping_num_leds;
 
+#    ifdef RGBLIGHT_LAYERS
+    // Write any enabled LED layers into the buffer
+    if (rgblight_layers != NULL) {
+        uint8_t i = 0;
+        // For each layer
+        for (const rgblight_segment_t * const *layer_ptr = rgblight_layers; pgm_read_ptr(layer_ptr) != NULL && i < RGBLIGHT_MAX_LAYERS; layer_ptr++, i++) {
+            if (rgblight_get_layer_state(i)) {
+                rgblight_segment_t segment;
+                // For each segment
+                for (const rgblight_segment_t * segment_ptr = pgm_read_ptr(layer_ptr);
```

Comment:
What a shame to call pgm_read_ptr() twice...

Reply:
Indeed! Restructured the code to only call it once.